python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -154,172 +155,93 @@ def _get_latest_materialization_record(
         )
         return next(iter(records), None)
 
-    def materialization_exists(
+    @cached_method
+    def _get_latest_storage_ids(
+        self, *, asset_key: AssetKey
+    ) -> Mapping[AssetKeyPartitionKey, Optional[int]]:
+        """Returns a mapping from asset partition to the latest storage id for that asset partition
+        for all asset partitions associated with the given asset key.
+
+        Note that for partitioned assets, an asset partition with a None partition key will be
+        present in the mapping, representing the latest storage id for the asset as a whole.
+        """
+        asset_partition = AssetKeyPartitionKey(asset_key)
+        latest_record = self._get_latest_record(asset_partition=asset_partition)
+        latest_storage_ids = {
+            asset_partition: latest_record.storage_id if latest_record is not None else None
+        }
+        if self.asset_graph.is_partitioned(asset_key):
+            latest_storage_ids.update(
+                {
+                    AssetKeyPartitionKey(asset_key, partition_key): storage_id
+                    for partition_key, storage_id in self.instance.get_latest_storage_id_by_partition(
+                        asset_key, event_type=self._event_type_for_key(asset_key)
+                    ).items()
+                }
+            )
+        return latest_storage_ids
+
+    def get_latest_storage_id(self, asset_partition: AssetKeyPartitionKey) -> Optional[int]:
+        """Returns the latest storage id for the given asset partition. If the asset has never been
+        materialized, returns None.
+
+        Args:
+            asset_partition (AssetKeyPartitionKey): The asset partition to query.
+        """
+        return self._get_latest_storage_ids(asset_key=asset_partition.asset_key).get(
+            asset_partition
+        )
+
+    def record_exists(
```

Comment:
I think this needs a more specific name. `asset_partition_has_observation_or_materialization`?